Windows conda build

StephenSmith25 opened the pull request:
**Description of work.**
This PR adds the conda windows environment. 


<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

The following instructions will create a build in a source tree build folder e.g mantid/build. To specify another build directory please pass -B to cmake.

1. Install miniconda https://repo.anaconda.com/miniconda/Miniconda3-latest-Windows-x86_64.exe
2. Open an Anaconda Prompt (Miniconda 3)
3. Go to your mantid source folder
3. mkdir build
4. cd build
5. conda env create -f ../mantid-developer-win.yml
6. conda activate mantid-developer
7. cmake .. --preset= win
8. Open visual studio with visual-studio.bat
9. Change build type to RelWithDebInfo or Release (Debug builds will not work)
10. Build everything
11. Set startup project as workbench
12. Launch workbench

Optionally, steps 8-12 can be replaced with

1.  cmake --build . --config RelWithDebInfo
2. thirdpartypaths.bat
3. cd bin/RelWithDebInfo
4. cmd /c "set PYTHONPATH=%PYTHONPATH%;.&workbench.exe"
(last step seems quite verbose, maybe I've missed something)






<!-- Instructions for testing. -->

Fixes #31614. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

StephenSmith25 committed:
Add cmake support for windows conda env

StephenSmith25 committed:
Fix conda windows precommit

StephenSmith25 committed:
Add cmake preset and make env file consistent with OSX and Linux

StephenSmith25 committed:
Amend CXX path info for conda

StephenSmith25 committed:
Slight tolerance change in SANSFitShiftScaleTest

StephenSmith25 committed:
Code changes

StephenSmith25 committed:
Fix FindNexus for master windows build

StephenSmith25 committed:
Amend hdf5 find_package call

StephenSmith25 committed:
Disable bayes libraries on conda

StephenSmith25 committed:
Fix lib3mf finder

StephenSmith25 committed:
Add conda launch scripts for visual studio and pycharm

StephenSmith25 committed:
Amend msvc debugger environment

StephenSmith25 committed:
Revert changes to workbench VS target

StephenSmith25 committed:
Add developer suffix to environment

DannyHindson reviewed (changes requested):
I went through the test instructions and the set up went OK. I found a couple of problems with the build and running of workbench though.

I got these compiler warnings from the build:

![image](https://user-images.githubusercontent.com/56295817/124431541-640f6580-dd68-11eb-9b60-109555a146c7.png)

...and going to this unit\line I can see that the Visual Studio intellisense isn't v happy about this constant:

![image](https://user-images.githubusercontent.com/56295817/124431612-77bacc00-dd68-11eb-8e4a-1568a6267491.png)

I think it's perhaps a forward slash\back slash problem. If I change this to the path that's commented out the warnings go away.

I think this issue also stops workbench starting up OK. I was getting a boost process error but when I changed the code to the commented out path in my screenshot above it started up OK.

I also got this compiler error - but haven't seen any negative side effects of this yet. I guess it would stop one of the tests running:

![image](https://user-images.githubusercontent.com/56295817/124431767-b5b7f000-dd68-11eb-9499-cbe0ed21b1c6.png)

DannyHindson reviewed (changes requested):
Couple of other documentation related thing.

When I try and access the qt documentation from the workbench help menu, workbench crashes. I get the "Python has stopped working" window and then in the console window I get this:

![image](https://user-images.githubusercontent.com/56295817/124443103-4e546d00-dd75-11eb-9f1b-f9e26c2a85d2.png)

Not sure if it's related but I got these errors when building the docs-qtassistant target:

![image](https://user-images.githubusercontent.com/56295817/124443241-6c21d200-dd75-11eb-882f-b8148ae2dc59.png)

StephenSmith25 committed:
Fix mantidworkbench.cpp python path

StephenSmith25 committed:
Fix qt help docs not launching

StephenSmith25 commented:
Thanks for giving this a test! 

Latest commits should fix the two issues you found. With regards to the docs warnings, I got those errors about unicode as well and then rebuilt the docs target and they disappeared 🤷‍♂️. Not sure if the errors are a conda issue or a mantid docs issue.

DannyHindson commented:
Thanks for the fixes. Workbench starts up fine now and I can also open the docs. Re the docs warnings I had a look at an example .rst file where there is one of the characters reported:

docs\source\tutorials\extending_mantid_with_python\advanced_fit_behaviours\05_exercise_6.rst

This has a Gamma character in it which is U+0393. If I visit the page from workbench it's not displaying properly in either the Conda build or the nightly:

![image](https://user-images.githubusercontent.com/56295817/124503491-99926e00-ddbd-11eb-85e5-fbb822b46b4d.png)

So not related to this PR.

I've found one other problem which is that I can't start up workbench from PyCharm. When I try it I get this message:

![image](https://user-images.githubusercontent.com/56295817/124503827-381ecf00-ddbe-11eb-960e-57ac09e7b7e2.png)

StephenSmith25 commented:
So to get pycharm working I did the following:

- Run pycharm with pycharm.bat in the build folder (this appears to be optional)
- Create a new project with a conda python interpreter (the mantid-developer one) 
- Setup the paths as per https://developer.mantidproject.org/GettingStartedWithPyCharm.html
- Create a workbench configuration as shown
![image](https://user-images.githubusercontent.com/56431339/124572227-65ac5c80-de40-11eb-8faf-566aad6f0731.png)

- Hopefully it should launch

DannyHindson commented:
Thanks. That's fixed it. I can run workbench from PyCharm now

So the only issue I've got left is that compiler error from the PythonInterfaceCppTest. I wonder if this relates to the version of the MSVC compiler being used. I'm running Visual Studio 16.8.3 - does this test build OK for you? I see that the test did build OK on Jenkins so maybe not the end of the world.

StephenSmith25 committed:
add compile definition to PythonInterfaceCppTest

This definition is required on windows as the python includes from boost redefine snprintf (as this was necessary for older versions of MSVC). This now breaks things when you build with a newer version of MSVC.

StephenSmith25 commented:
I was getting that error as well, I hadn't noticed it before (maybe it’s a new test). It turns out there’s an issue with boost headers (in >v1.69) redefining snprint for some compatibility reasons. In visual studio > 15 you need to add a compile definition to stop this redefinition as MSVC have solved this issue. 

See https://github.com/boostorg/system/issues/32

DannyHindson commented:
> I was getting that error as well, I hadn't noticed it before (maybe it’s a new test). It turns out there’s an issue with boost headers (in >v1.69) redefining snprint for some compatibility reasons. In visual studio > 15 you need to add a compile definition to stop this redefinition as MSVC have solved this issue.
> 
> See [boostorg/system#32](https://github.com/boostorg/system/issues/32)

Thanks that's fixed it

DannyHindson commented on scripts/Inelastic/IndirectImport.py:
```diff
@@ -80,7 +81,7 @@ def is_supported_f2py_platform():
     libraries, else False.
     """
-    if (_os_env().startswith("Windows")
+    if (_os_env().startswith("Windows") and "CONDA_PREFIX" not in os.environ
```
Can I just double check this? I thought one of the benefits of the Conda approach was that we would be able to support f2py across all the platforms. This seems to exclude it though for any Conda build

StephenSmith25 replied:
That will be the intention, but I was in two minds of whether to include it in this PR so for the purposes of testing I turned it off. I added the f2py modules through the PR https://github.com/mantidproject/mantid/pull/31862 . There's still some thought required on how to integrating these libraries, e.g. adding the relevant things to the python search path. 

As I understand it, the libraries used in `IndirectImport` assume that we'll use the libraries in that directory, e.g `QLdata_win64.cp38-win_amd64.pyd`. These libraries are linked against a certain gfortran math library, which is shipped in our third-party dependencies https://github.com/mantidproject/thirdparty-msvc2015/tree/master/bin/mingw . Without these dependencies (as we don't download the third-party repo), the libraries have unmet dependencies. Therefore, the current setup won't work with conda, so a bit more thought needs to go into it.

Admittedly this isn't a great solution, but hopefully it will only be a temporary change, while we plan our f2py strategy going forward.

DannyHindson reviewed (approved):
Looks good. I tested out the lib3mf stuff as well and the .3mf file loading in LoadSampleEnvironment still works fine
